Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp

```diff
@@ -284,29 +285,33 @@ void RemoteGraphicsContextGL::copyTextureFromVideoFrame(WebKit::RemoteVideoFrame
 }
 #endif
 
-void RemoteGraphicsContextGL::simulateEventForTesting(WebCore::GraphicsContextGL::SimulatedEventForTesting event)
+void RemoteGraphicsContextGL::simulateEventForTesting(WebCore::GraphicsContextGL::SimulatedEventForTesting event, CompletionHandler<void()>&& completionHandler)
 {
     assertIsCurrent(workQueue());
+    switch (event) {
     // FIXME: only run this in testing mode. https://bugs.webkit.org/show_bug.cgi?id=222544
-    if (event == WebCore::GraphicsContextGL::SimulatedEventForTesting::Timeout) {
-        // Simulate the timeout by just discarding the context. The subsequent messages act like
-        // unauthorized or old messages from Web process, they are skipped.
-        callOnMainRunLoop([gpuConnectionToWebProcess = m_gpuConnectionToWebProcess, identifier = m_graphicsContextGLIdentifier]() {
+    case WebCore::GraphicsContextGL::SimulatedEventForTesting::Timeout:
+        // Simulate the timeout by just discarding the context. The subsequent
+        // messages act like unauthorized or old messages from Web process, they
+        // are skipped.
+        callOnMainRunLoopAndWait([gpuConnectionToWebProcess = m_gpuConnectionToWebProcess, identifier = m_graphicsContextGLIdentifier]() {
             if (auto connectionToWeb = gpuConnectionToWebProcess.get())
                 connectionToWeb->releaseGraphicsContextGLForTesting(identifier);
         });
-        return;
-    }
-    if (event == WebCore::GraphicsContextGL::SimulatedEventForTesting::ContextChange) {
+        break;
+    case WebCore::GraphicsContextGL::SimulatedEventForTesting::ContextChange:
 #if PLATFORM(MAC)
         callOnMainRunLoop([weakConnection = m_gpuConnectionToWebProcess]() {
             if (auto connection = weakConnection.get())
                 connection->dispatchDisplayWasReconfiguredForTesting();
         });
 #endif
-        return;
+        break;
+    default:
+        m_context->simulateEventForTesting(event);
+        break;
     }
-    m_context->simulateEventForTesting(event);
+    completionHandler();
```

Comment:
Yeah, this is a nice simple solution to the problem.. it may create problems down the line, but let's go with this one for now..

Technically it can still race executing the commands in the buffer. 
- GPUP processes the message buffer. It sees simulateEventForTesting.
- GPUP executes simulateEventForTesting. This removes the stream connection from the work queue. This does not stop the message buffer processing.
- WP sees simulateEventForTesting reply. WP submits the 3 commands after simulateEventForTesting finishes.
- GPUP still processes the message buffer and sees new commands and processes those and their replies.

I think we'll need to implement `StreamServerConnection::invalidate()` that immediately stops the buffer processing. However, that creates the problem "down the line", where the reply by `completionHandler()` will not be delivered at that point. But maybe when that's implemented, we just switch this message back to async.

Reply:
In local testing, I'm still seeing the test timeout even after making all the calls sync.

Back to the drawing board.